src/devices/src/virtio/queue.rs

```diff
@@ -140,7 +143,7 @@ impl<'a> DescriptorChain<'a> {
     }
 }
 
-#[derive(Clone)]
+#[derive(Clone, Debug, PartialEq, Versionize)]
```

Comment:
This is pending the decision on how we want to create the state for devices in general, but just wanted to lay out some disadvantages to this approach on queues.

The problem at hand here is that we are adding `Versionize` to the most common virtio device building block. This is a component that we want to consume from rust-vmm and it would be far easier to do so if the `Versonize` can be hidden behind a feature. For example, if we would have `Versonize` on `QueueState` instead of on `Queue`, the virtio crate could have a feature called `versioned_serialization`.

In the virtio `Cargo.toml` file:
```rust
....

# This way the dependencies for snapshot are only pulled when
# the feature `versioned_serialization` is used.

snapshot = { version = "0.1.0", optional = true }
versionize = { version = "0.1.0", optional = true }
versionize_derive = { version = "0.1.0", optional = true }

[features]
versioned_serialization = ["snapshot", "versionize", "versionize_derive"]
```

The `Queue` implementation stays the same irrespective of the feature `versioned_serialization` being specified or not. And the `QueueState` would be gated by the `versioned_serialization` feature as follows:

```rust
#[cfg(feature = "versioned_serialization")]
#[derive(Clone, Debug, Versionize)]
pub struct QueueState {
    max_size: u16,
    pub size: u16,
    pub ready: bool,
    pub desc_table: u64,
    pub avail_ring: u64,
    pub used_ring: u64,
    next_avail: Wrapping<u16>,
    next_used: Wrapping<u16>,
}
```
We don't need to have features in Firecracker of course, but I believe this approach would make it easier to consume the virtio building blocks from rust-vmm and it offers a cleaner separation of concerns, the concerns here being the virtio device functionality and the versioned serialization. We should also consider that `Queue` and `QueueState` are almost the same thing only now. If we add split queues or other features, this might not be the case any more.

Reply:
I am inclined to not duplicate all this state since `Queue` is a stable object, and is completely comprised of plain-old-data.

@andreeaflorescu I think we can achieve what you want in rust-vmm having a `versioned_serialization` feature with the current code as well by deriving `Versionize` only when the feature is present:

```rust
#[derive(Debug, Clone)]
#[cfg_attr(feature = "versioned_serialization", derive(Versionize))]
pub struct Queue { ... }
```

without duplicating and maintaining a shadow state structure.

Reply:
Yap, that's true. I like it more completely separated because you only need to use features at the `lib.rs` level so it should be easier to maintain. This is a soft requirement though.

We can leave it as it is in this PR, and only change it in the future in this case:
> We should also consider that Queue and QueueState are almost the same thing only now. If we add split queues or other features, this might not be the case any more.

The queue state can be the same irrespective on the type of queue (split/packed). But we might add some helper fields depending on the queue type. When it comes to that, we can easily split the structures in two (Queue and QueueState).

Reply:
It's not clear which is the better long-term option to me either. Let's discuss it live in our sync-up meeting.

Reply:
We decided that it's best to have a separate `QueueState`. This allows possible future changes to `Queue` that will not affect the state and also has the benefit of not having to versionize the `GuestAddress`.